packages/loopring_v3/contracts/converters/BaseConverter.sol

```diff
@@ -0,0 +1,202 @@
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event ConversionSuccess (uint amountIn, uint amountOut);
+    event ConversionFailed  (string reason);
+
+    IExchangeV3        public  immutable exchange;
+    IDepositContract   public  immutable depositContract;
+
+    bool               public  initialized;
+
+    address            public  tokenIn;
+    address            public  tokenOut;
+
+    bool               public  failed;
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchange.owner(), "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3 _exchange
+        )
+    {
+        exchange = _exchange;
+        depositContract = _exchange.getDepositContract();
+    }
+
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        uint8         _decimals,
+        address       _tokenIn,
+        address       _tokenOut
+        )
+        external
+    {
+        require(!initialized, "ALREADY_INITIALIZED");
+        initializeToken(_name, _symbol, _decimals);
+
+        tokenIn = _tokenIn;
+        tokenOut = _tokenOut;
+
+        initialized = true;
+    }
+
+    function deposit(
+        uint96          amountIn,
+        uint96          minAmountOut,
+        bytes  calldata customData
+        )
+        external
+        payable
+        onlyFromExchangeOwner
+    {
+        require(totalSupply == 0);
+
+        // Converter specific logic, which can fail
+        try BaseConverter(this).convertExternal(amountIn, minAmountOut, customData)
+            returns (uint amountOut) {
+            failed = false;
+            emit ConversionSuccess(amountIn, amountOut);
+        } catch Error(string memory reason) {
+            failed = true;
+            emit ConversionFailed(reason);
+        } catch {
+            failed = true;
+            emit ConversionFailed("unknown");
+        }
+
+        // Mint pool tokens representing each user's share in the pool, with 1:1 ratio
+        _mint(address(this), amountIn);
+
+        // Repay the flash mint used to give user's their share on L2
+        _repay(address(this), amountIn);
+    }
+
+    function withdraw(
+        address to,
+        uint96  poolAmount,
+        uint96  repayAmount
+        )
+        public
+    {
+        // Token to withdraw
+        address token = failed ? tokenIn : tokenOut;
+
+        // Current balance in the contract
+        uint balance = 0;
+        if (token == address(0)) {
+            balance = address(this).balance;
+        } else {
+            balance = ERC20(token).balanceOf(address(this));
+        }
+
+        // Share to withdraw
+        uint amount = balance.mul(poolAmount) / totalSupply;
+
+        // Burn pool tokens
+        _burn(msg.sender, poolAmount);
+
+        // Use to repay flash mint directly if requested
+        if (repayAmount > 0) {
+            _repay(token, repayAmount);
```

Comment:
I see the idea is: for the second phase, we still use flash deposit, then use this `withdraw` method to pay either Token or yToken to trade with users' vToken. 

But since we already have the real token (Token or yToken) inside the Converter contract, an alternative is that the Converter does a real deposit of either Token or yToken before the next block is submitted using the same Ethereum tx. Therefore, the second phase doesn't need flash mint/deposit. 

I think that will be more efficient.

Reply:
> I see the idea is: for the second phase, we still use flash deposit, then use this withdraw method to pay either Token or yToken to trade with users' vToken.

Yep.

> But since we already have the real token (Token or yToken) inside the Converter contract, an alternative is that the Converter does a real deposit of either Token or yToken before the next block is submitted using the same Ethereum tx. Therefore, the second phase doesn't need flash mint/deposit.

The tokens are indeed already n the converter contract, but we cannot access them! The tokens in the vault are fully owned by the owners of the vault tokens, and we gave those to the users on L2 in phase 1 in return for the tokens they want to sell. And it's important we give the vault tokens to the users, otherwise they would stop being the owner of their funds in the converter and someone else/we could just take them. Only in return for the vault tokens are the tokens in the converter unlocked, so we first need to to collect them on L2.

Reply:
I see the point.